qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,647 @@
+    if several control channels work together on the same qubits. For example, if the
+    control channel configuration is {(3,2): [ControlChannel(3), ControlChannel(12)]}
+    then given qubits (2, 3) the name "ch1.0$1" will resolve to ControlChannel(12) while
+    "ch1.0$0" will resolve to ControlChannel(3).
+    """
+
+    def __init__(self, control_config: Dict[Tuple[int, ...], List[ControlChannel]] = None):
+        """
+        Initialize the instructions from a given backend.
+
+        Args:
+            control_config: A configuration dictionary of any control channels. The
+                keys are tuples of qubits and the values are a list of ControlChannels
+                that correspond to the qubits in the keys.
+        """
+
+        self._controls_config = control_config if control_config else {}
+
+        # Dict of the form: (schedule.name, parameter.name, qubits): Parameter
+        self._parameter_map = {}
+
+        # Reverse mapping of _parameter_map
+        self._parameter_map_r = defaultdict(set)
+
+        # Default dict of the form: (schedule.name, parameter.name, qubits): [ParameterValue, ...]
+        self._params = defaultdict(list)
+
+        self._schedules = {}
+
+        self._channel_pattern = "^ch\d[.\d]*\${0,1}[\d]*$"
+
+    def add_schedule(self, schedule: Schedule, qubits: Tuple = None):
+        """
+        Add a schedule and register its parameters.
+
+        Args:
+            schedule: The schedule to add.
+            qubits: The qubits for which to add the schedules. If None is given then this
+                schedule is the default schedule for all qubits.
+
+        Raises:
+            CalibrationError: If the parameterized channel index is not formatted
+                following index1.index2... or if several parameters in the same schedule
+                have the same name.
+        """
+        # check that channels, if parameterized, have the proper name format.
+        # pylint: disable = raise-missing-from
+        param_indices = set()
+        for ch in schedule.channels:
+            if isinstance(ch.index, Parameter):
+                param_indices.add(ch.index)
+                if re.compile(self._channel_pattern).match(ch.index.name) is None:
+                    raise CalibrationError(
+                        f"Parameterized channel must correspond to {self._channel_pattern}"
+                    )
+
+        self._schedules[(schedule.name, qubits)] = schedule
+
+        param_names = [param.name for param in schedule.parameters]
+
+        if len(param_names) != len(set(param_names)):
+            raise CalibrationError(f"Parameter names in {schedule.name} must be unique.")
+
+        # Register parameters that are not indices.
+        for param in schedule.parameters:
+            if param not in param_indices:
+                self.register_parameter(param, schedule, qubits)
+
+    def register_parameter(
```

Comment:
Is this necessary to be a public method? Because `Schedule` contains parameter to be added, an end-user separately specifies a parameter looks strange to me.

Reply:
I think you are right. There could be a case where a user would want to register a parameter without a schedule but in that case I would have a wrapper method. For now I have made this registration method private: fa27ef5